liteflow-labs/liteflow-js
# Implement js function in compiler

### Work done
- Implemented the js function task in the compiler.
- Renamed in the process compiler `instanceHash` by `instanceId`. Also, change the type to UUID.
- Fixed in the process compiler the output of instance. Now only `instance` OR `instanceId` are sent to the API.
- Put the preProcess functions as async.
- Deployed the js-function on prod backend and use its instance id in the CLI.

### How to test
The task is using the `handler` parameter to specify the js file to execute:

```yaml
name: function
steps:
  - type: trigger
    endpointKey: function
  - type: task
    handler: ./foo.js
    inputs:
      key: body
```

with `foo.js`:
```js
module.exports = (inputs) => {
  if (inputs.a === undefined || isNaN(inputs.a))
    throw new Error("input a is not defined or is not a number")
  if (inputs.b === undefined || isNaN(inputs.b))
    throw new Error("input b is not defined or is not a number")
  return parseFloat(inputs.a) + parseFloat(inputs.b)
}
```

Make sure the file can be run by Node v12 with a simple `node foo.js`. If no error is returned, then Node can interpret the js code.

Deploy the app like any other app.
Execute the process with:
```
curl -XPOST -H "Content-Type: application/json" --data '{"a":2,"b":"30"}' https://gateway.liteflow.com/XXXX

# to make the function throw an error:
curl -XPOST -H "Content-Type: application/json" --data '{"a":2,"b":"foo"}' https://gateway.liteflow.com/XXXX
```

### Future todo (to move to issues)
- [ ] use vm2 with all default module except fs in js service
- [ ] add top 100 npm modules
- [ ] upload source online instead of the task
- [ ] allow user to run a command to build the source before uploading them
